packages/template/-private/dsl/emit.d.ts

```diff
@@ -64,16 +64,38 @@ export declare function emitComponent<T extends AcceptsBlocks<any, any>>(
   blockParams: T extends AcceptsBlocks<infer Yields, any> ? Required<Yields> : any;
 };
 
-/**
- * Acts as a top-level wrapper for translated template bodies. The given
- * callback accepts a template context value as well as an instance of the
+/*
+ * Wraps a template body that appears as a standalone expression and is therefore not
+ * associated with any backing value.
+ *
+ * The given callback accepts a template context value as well as an instance of the
  * environment's DSL export.
  */
-export declare function template<
+export declare function templateExpression<
   Signature extends AnyFunction = (args: EmptyObject) => AcceptsBlocks<EmptyObject>,
   Context extends AnyContext = TemplateContext<void, EmptyObject, EmptyObject, void>
 >(f: (𝚪: Context, χ: never) => void): new () => Invokable<Signature> & HasContext<Context>;
 
+/*
+ * Wraps a template body that's backed by a known value (typically a class), either
+ * via a `.hbs` association to a default export or via embedding e.g. with `<template>`.
+ *
+ * The given callback accepts a template context value as well as an instance of the
+ * environment's DSL export.
+ *
+ * Note that the constructor-based signature returns an (arbitrary) abstract constructor
+ * type in order to trigger TypeScript's higher-order function type inference and propagate
+ * any type parameters on the class (see https://github.com/microsoft/TypeScript/pull/30215)
```

Comment:
I don't doubt this makes perfect sense, but I have a few questions such that it doesn't actually make sense *to me* right now:

1. What is the `void` return value for (vs. the constructor version)?

2. Similarly, what is the version which *doesn't* have `Args` passed into it doing? Are these `Args` a `Signature['Args']` or some other `Args`? I *think* the latter, but the prominence of `Signature['Args']` makes me doubt that; in any case I don't know how they relate!

Net… what is each overload here supplying? 😬

Reply:
> What is the void return value for (vs. the constructor version)?
>
> Similarly, what is the version which doesn't have Args passed into it doing?

Similar to the constructor return type, the `Args` aren't supplying anything; it's just necessary to have a type of that shape/position in order to trigger higher-order function type inference. Without that, TS won't infer a function type with type parameters for `body`, which is the root cause of the error showing up in #389.

This playground demonstrates the issue: https://tsplay.dev/w1pAyW

> Net… what is each overload here supplying? 😬

The overloads here are a 1:1 mapping from the original `ResolveContext` type:
```ts
export type ResolveContext<T> = T extends HasContext<infer Context>
  ? Context
  : T extends Constructor<HasContext<infer Context>>
  ? Context
  : unknown;
```

One case handles receiving a constructor (which is the common case), and the other handles non-constructor values (which `TemplateOnlyComponent` was originally an example of, though we might have made that a constructor today; I don't remember off the top of my head).

Reply:
> I wonder if we still need to keep non-constructor value support here, though, given that TemplateOnlyComponents now by definition have to go through the other path?

That's a good callout. I left it because I was uncertain and just directly translating things over, but it's easy enough to drop the second overload and see if anything breaks. I'll try that and report back.

Reply:
Looks like it works fine for us. Thanks for pointing it out!

Reply:
Sweet!